### DISCUSSION TITLE
Proper HEAD support by having non-graphical components

### DISCUSSION THREAD

**Discussion by SamMousa** [GrapesJS/grapesjs#6120] [Roadmap Suggestions]
I read in some issues (from 2018) that GrapesJS by default ignores anything outside body. The argument was that it is a page editor not a metadata editor.
Currently there is some code that gives partial support for full documents or other workarounds:
1. `asDocument` import does some saving of the `head` content.
2. There is code in GrapesJS that moves styles from the `head` to the `body` so it works properly.

Currently components exist in the wild that model non-visible contents. Like a component for google tag manager (not a fan, just using it as an example).

## Proposal

Use this as a base for discussion, it has holes and is not even 90% thought through.

What about just iterating over elements in the `<head>` just like elements in the body.
Instead of rendering the components to the page canvas we just create an extra canvas (like a `metadata` tab in the editor).
This is where these components could then render their UI, again via the normal API.
The main canvas would then render the head components' HTML.

This would enable interesting usages like using an image picker to pick a favicon. Or using a `LinkTagBlock` to include a stylesheet.

From a framework perspective not a lot of changes are required as far as I can tell; the component API needs little to no changes.
In "userland" one could even implement this, simply but ugly:
1. Split head content from the HTML.
2. Create new HTML with just the HEAD content inside a BODY.
3. Create a GrapesJS instance (A) that uses the XML parser; configure this to just have blocks for a HEAD tag.
4. Create a GrapesJS instance (B) that parses the body from 1 and is our standard editor.
5. When making changes in A render the HTML from the components and put it in the HEAD of the canvas in B.

**Comment by artf** [comment]
I think I got the idea but I'm not sure about the addition you're suggesting.

I mean, technically if you want to render the head and its elements + selecting and updating them, I'd expect it already to be possible. You should extend the view of those components and make them "displayable" (browsers hide them by default).

**Reply by SamMousa** [reply]
When using the HTML parser the head is not passed to custom blocks though. (I'm logging every element on passed to my `isComponent(..)` callback. 
Possibly switching to the XML parser could work, but it's not entirely clear what the risks of this are.
Furthermore even if I do that, inside the canvas those elements are then also rendered inside the `<head>` which means they'll never be visible.

Instead we should have a canvas for components inside the head of the primary document, which are rendered in the body of the secondary canvas. The primary canvas then doesn't render the components, but only includes the resulting HTML.

**Reply by artf** [reply]
You just need to extend the head.
https://jsfiddle.net/artur_arseniev/6r4by83q
